File: android/src/main/java/com/nativeguard/WireGuardVpnModule.java
package com.nativeguard;

import com.facebook.react.bridge.*;
import com.facebook.react.module.annotations.ReactModule;
import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

@ReactModule(name = WireGuardVpnModule.NAME)
public class WireGuardVpnModule extends ReactContextBaseJavaModule {
    public static final String NAME = "WireGuardVpnModule";

    public WireGuardVpnModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @NonNull
    @Override
    public String getName() {
        return NAME;
    }

    /**
     * Initialize the VPN backend (e.g. permissions, service setup).
     */
    @ReactMethod
    public void initialize(Promise promise) {
        try {
            // TODO: actual initialization logic
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("INITIALIZE_ERROR", "Failed to initialize VPN backend", e);
        }
    }

    /**
     * Start a WireGuard tunnel with the given config.
     */
    @ReactMethod
    public void connect(ReadableMap config, Promise promise) {
        try {
            // Validate required fields
            if (!config.hasKey("privateKey") || config.getString("privateKey") == null) {
                promise.reject("INVALID_CONFIG", "Missing 'privateKey'");
                return;
            }
            if (!config.hasKey("publicKey") || config.getString("publicKey") == null) {
                promise.reject("INVALID_CONFIG", "Missing 'publicKey'");
                return;
            }
            if (!config.hasKey("serverAddress") || config.getString("serverAddress") == null) {
                promise.reject("INVALID_CONFIG", "Missing 'serverAddress'");
                return;
            }
            if (!config.hasKey("serverPort")) {
                promise.reject("INVALID_CONFIG", "Missing 'serverPort'");
                return;
            }

            // Extract fields
            String privateKey    = config.getString("privateKey");
            String publicKey     = config.getString("publicKey");
            String serverAddress = config.getString("serverAddress");
            int    serverPort    = config.getInt("serverPort");

            // allowedIPs
            ReadableArray allowedIPsArray = config.getArray("allowedIPs");
            if (allowedIPsArray == null) {
                promise.reject("INVALID_CONFIG", "Missing 'allowedIPs'");
                return;
            }
            List<String> allowedIPs = new ArrayList<>();
            for (int i = 0; i < allowedIPsArray.size(); i++) {
                allowedIPs.add(allowedIPsArray.getString(i));
            }

            // dns
            ReadableArray dnsArray = config.getArray("dns");
            if (dnsArray == null) {
                promise.reject("INVALID_CONFIG", "Missing 'dns'");
                return;
            }
            List<String> dns = new ArrayList<>();
            for (int i = 0; i < dnsArray.size(); i++) {
                dns.add(dnsArray.getString(i));
            }

            // optional fields
            int    mtu           = config.hasKey("mtu") ? config.getInt("mtu") : 1420;
            String presharedKey  = config.hasKey("presharedKey") 
                                    ? config.getString("presharedKey") 
                                    : null;

            // TODO: use these values to start the WireGuard tunnel
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("CONNECT_ERROR", "Error connecting to VPN: " + e.getMessage(), e);
        }
    }

    /**
     * Tear down the current tunnel.
     */
    @ReactMethod
    public void disconnect(Promise promise) {
        try {
            // TODO: stop the VPN tunnel
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("DISCONNECT_ERROR", "Error disconnecting VPN: " + e.getMessage(), e);
        }
    }

    /**
     * Return one of the WireGuardStatus strings.
     */
    @ReactMethod
    public void getStatus(Promise promise) {
        try {
            // TODO: query real tunnel state
            String status = "disconnected";
            promise.resolve(status);
        } catch (Exception e) {
            promise.reject("STATUS_ERROR", "Error fetching VPN status: " + e.getMessage(), e);
        }
    }

    /**
     * Check whether this device/platform supports WireGuard.
     */
    @ReactMethod
    public void isDeviceSupported(Promise promise) {
        try {
            // TODO: check kernel version or required APIs
            promise.resolve(true);
        } catch (Exception e) {
            promise.reject("SUPPORTED_ERROR", "Error checking support: " + e.getMessage(), e);
        }
    }
}

File: android/src/main/java/com/compiledbox/nativeguard/WireGuardVpnModule.kt
package com.compiledbox.nativeguard

import android.app.Activity
import android.content.Intent
import android.net.VpnService
import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule

@ReactModule(name = WireGuardVpnModule.NAME)
class WireGuardVpnModule(private val ctx: ReactApplicationContext) :
    ReactContextBaseJavaModule(ctx), ActivityEventListener {

    companion object {
        const val NAME = "WireGuardVpnModule"
        private const val PREPARE_VPN = 0xABCD
    }

    private var initPromise: Promise? = null

    init {
        ctx.addActivityEventListener(this)
    }

    override fun getName(): String = NAME

    /** Request user permission for VPN. */
    @ReactMethod
    fun initialize(promise: Promise) {
        val activity = currentActivity
        if (activity == null) {
            promise.reject("INITIALIZE_ERROR", "No current activity")
            return
        }
        val intent = VpnService.prepare(ctx)
        if (intent != null) {
            initPromise = promise
            activity.startActivityForResult(intent, PREPARE_VPN)
        } else {
            // Already granted
            promise.resolve(null)
        }
    }

    override fun onActivityResult(
        activity: Activity,
        requestCode: Int,
        resultCode: Int,
        data: Intent?
    ) {
        if (requestCode == PREPARE_VPN) {
            initPromise?.apply {
                if (resultCode == Activity.RESULT_OK) resolve(null)
                else reject("INITIALIZE_ERROR", "VPN permission denied")
            }
            initPromise = null
        }
    }

    override fun onNewIntent(intent: Intent?) = Unit

    /** Build config, start the service, and bring up the tunnel. */
    @ReactMethod
    fun connect(configMap: ReadableMap, promise: Promise) {
        try {
            val configText = buildConfigText(configMap)
            // Start the service in foreground
            val intent = Intent(ctx, WireGuardVpnService::class.java)
            if (android.os.Build.VERSION.SDK_INT >= 26)
                ctx.startForegroundService(intent)
            else
                ctx.startService(intent)
            // Give the service a moment to initialize
            Thread.sleep(200)
            WireGuardVpnService.instance?.startTunnel("nativeguard", configText)
                ?: throw IllegalStateException("VPN service not running")
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("CONNECT_ERROR", "Error connecting VPN: ${e.message}", e)
        }
    }

    /** Stop the service’s tunnel. */
    @ReactMethod
    fun disconnect(promise: Promise) {
        try {
            WireGuardVpnService.instance?.stopTunnel()
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("DISCONNECT_ERROR", "Error disconnecting VPN: ${e.message}", e)
        }
    }

    /** Query current state from the service. */
    @ReactMethod
    fun getStatus(promise: Promise) {
        try {
            val status = WireGuardVpnService.instance?.getTunnelState() ?: "disconnected"
            promise.resolve(status)
        } catch (e: Exception) {
            promise.reject("STATUS_ERROR", "Error getting status: ${e.message}", e)
        }
    }

    /** Check API level for VPNService support. */
    @ReactMethod
    fun isDeviceSupported(promise: Promise) {
        promise.resolve(android.os.Build.VERSION.SDK_INT >= 21)
    }

    /** Render a full WireGuard .conf from the JS-provided map. */
    private fun buildConfigText(cfg: ReadableMap): String {
        val priv = cfg.getString("privateKey")!!
        val pub = cfg.getString("publicKey")!!
        val addr = cfg.getString("serverAddress")!!
        val port = cfg.getInt("serverPort")
        val allowed = cfg.getArray("allowedIPs")!!.toArrayList().joinToString(",")
        val dns    = cfg.getArray("dns")!!.toArrayList().joinToString(",")
        val mtu    = if (cfg.hasKey("mtu")) cfg.getInt("mtu") else 1420
        val ps     = cfg.getString("presharedKey")

        return buildString {
            appendLine("[Interface]")
            appendLine("PrivateKey = $priv")
            appendLine("MTU = $mtu")
            appendLine()
            appendLine("[Peer]")
            appendLine("PublicKey = $pub")
            appendLine("Endpoint = $addr:$port")
            appendLine("AllowedIPs = $allowed")
            appendLine("DNS = $dns")
            ps?.let { appendLine("PresharedKey = $it") }
        }
    }
}
